kobo-uncaged/device/types.go

```diff
@@ -40,210 +38,79 @@ type firmwareVersion struct {
-	Timestamp       *string                `json:"timestamp" mapstructure:"timestamp"`
-	LastModified    *string                `json:"last_modified" mapstructure:"last_modified"`
-	UUID            string                 `json:"uuid" mapstructure:"uuid"`
-	TitleSort       string                 `json:"title_sort" mapstructure:"title_sort"`
-	AuthorLinkMap   map[string]string      `json:"author_link_map" mapstructure:"author_link_map"`
-	Title           string                 `json:"title" mapstructure:"title"`
-	Identifiers     map[string]string      `json:"identifiers" mapstructure:"identifiers"`
+// MetaIterator Kobo UNCaGED to lazy load book metadata
+type MetaIterator struct {
+	k        *Kobo
+	cidList  []string
+	cidIndex int
 }
 
-type koboDevice string
-
-// Kobo model ID's from https://github.com/geek1011/KoboStuff/blob/gh-pages/kobofirmware.js#L11
-const (
-	touchAB      koboDevice = "00000000-0000-0000-0000-000000000310"
-	touchC       koboDevice = "00000000-0000-0000-0000-000000000320"
-	mini         koboDevice = "00000000-0000-0000-0000-000000000340"
-	glo          koboDevice = "00000000-0000-0000-0000-000000000330"
-	gloHD        koboDevice = "00000000-0000-0000-0000-000000000371"
-	touch2       koboDevice = "00000000-0000-0000-0000-000000000372"
-	aura         koboDevice = "00000000-0000-0000-0000-000000000360"
-	auraHD       koboDevice = "00000000-0000-0000-0000-000000000350"
-	auraH2O      koboDevice = "00000000-0000-0000-0000-000000000370"
-	auraH2Oed2r1 koboDevice = "00000000-0000-0000-0000-000000000374"
-	auraH2Oed2r2 koboDevice = "00000000-0000-0000-0000-000000000378"
-	auraOne      koboDevice = "00000000-0000-0000-0000-000000000373"
-	auraOneLE    koboDevice = "00000000-0000-0000-0000-000000000381"
-	auraEd2r1    koboDevice = "00000000-0000-0000-0000-000000000375"
-	auraEd2r2    koboDevice = "00000000-0000-0000-0000-000000000379"
-	claraHD      koboDevice = "00000000-0000-0000-0000-000000000376"
-	forma        koboDevice = "00000000-0000-0000-0000-000000000377"
-	forma32gb    koboDevice = "00000000-0000-0000-0000-000000000380"
-	libra        koboDevice = "00000000-0000-0000-0000-000000000384"
-)
-
-// Model returns the model name for the device.
-func (d koboDevice) Model() string {
-	switch d {
-	case touch2, touchAB, touchC:
-		return "Touch"
-	case mini:
-		return "Mini"
-	case glo:
-		return "Glo"
-	case gloHD:
-		return "Glo HD"
-	case aura:
-		return "Aura"
-	case auraH2O:
-		return "Aura H2O"
-	case auraH2Oed2r1, auraH2Oed2r2:
-		return "Aura H2O Ed. 2"
-	case auraEd2r1, auraEd2r2:
-		return "Aura Ed. 2"
-	case auraHD:
-		return "Aura HD"
-	case auraOne, auraOneLE:
-		return "Aura One"
-	case claraHD:
-		return "Clara HD"
-	case forma, forma32gb:
-		return "Forma"
-	case libra:
-		return "Libra H2O"
-	default:
-		return "Unknown Kobo"
-	}
+// NewMetaIter creates a new MetaIterator for use
+func NewMetaIter(k *Kobo) *MetaIterator {
+	iter := MetaIterator{k: k, cidIndex: -1}
+	iter.cidList = make([]string, 0)
```

Comment:
This isn't required, as lists are empty by default (`[]string{} == make([]string, 0) == nil`).

```suggestion
```

Reply:
Are empty slices actually nil?

Reply:
Essentially. For all visible purposes, yes, they are the same. The only difference is in the underlying representation (see reflect.SliceHeader), in which the data pointer itself might not be nil (but it would still have a size/len of 0).

Reply:
Ah yeah, just did a quick playground test. Thanks for pointing it out.